ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/payloadattestation/MatchingDataPayloadAttestationGroup.java

```diff
@@ -61,37 +63,35 @@ int size() {
    * eventually be aggregated with others.
    */
   boolean add(final PayloadAttestationMessage payloadAttestationMessage) {
+    final int validatorIndex = payloadAttestationMessage.getValidatorIndex().intValue();
     if (!payloadAttestationMessage.getData().equals(data)) {
       // ignore payload attestation messages with a different data
       return false;
     }
-    return payloadAttestationMessages.add(payloadAttestationMessage);
+    return payloadAttestationMessages.putIfAbsent(validatorIndex, payloadAttestationMessage)
+        == null;
   }
 
   PayloadAttestation createAggregatedPayloadAttestation(final IntList ptc) {
     checkArgument(!payloadAttestationMessages.isEmpty(), "Nothing to aggregate");
     // use a snapshot to avoid any concurrent additions
-    final Set<PayloadAttestationMessage> payloadAttestationMessagesSnapshot =
-        new HashSet<>(payloadAttestationMessages);
-    final UInt64 slot = data.getSlot();
-    final int[] setBitIndices =
-        payloadAttestationMessagesSnapshot.stream()
-            .flatMapToInt(
-                message ->
-                    // relative position(s) of the validator index with respect to the PTC
-                    IntStream.range(0, ptc.size())
-                        .filter(i -> ptc.getInt(i) == message.getValidatorIndex().intValue()))
-            .toArray();
+    final Int2ObjectMap<PayloadAttestationMessage> payloadAttestationMessagesSnapshot =
+        new Int2ObjectOpenHashMap<>(payloadAttestationMessages);
+    // relative positions of the validator indices with respect to the PTC
+    final IntList setBitIndices = new IntArrayList();
+    final List<BLSSignature> signatures = new ArrayList<>();
+    for (int ptcPosition = 0; ptcPosition < ptc.size(); ptcPosition++) {
+      final int validatorIndex = ptc.getInt(ptcPosition);
+      if (payloadAttestationMessagesSnapshot.containsKey(validatorIndex)) {
+        setBitIndices.add(ptcPosition);
+        signatures.add(payloadAttestationMessagesSnapshot.get(validatorIndex).getSignature());
+      }
```

Comment:
do we actually need this snapshot?
if we do just a `payloadAttestationMessages.get` and handle the `null` case we can do a single pass and no consistency issues raise. Am I correct?